Feature/density chart

tdesfont opened the pull request.

tdesfont committed:
Starter code of working version
requirements.txt: List package versions
dku-helpers.js: Legacy helper from graph analytics code
HeatLayer.js/leaflet-heat.js: Embed leaflet code for the heat map
backend.py: Created main function for geospatial data

tdesfont committed:
Use graph analytics code base

tdesfont committed:
backend.py: Fix mistake

tdesfont committed:
app.js: Visualisation parametrization based on plugin user input
backend.py: Minor updates
webapp.json: Add or rename user inputs

tdesfont committed:
Minor wording

tdesfont committed:
Working mvp for the slider UI for intensity and radius

tdesfont committed:
Working state for the density chart and a single intensity html slider

tdesfont committed:
Include the event loop

tdesfont committed:
Include the event loop

tdesfont committed:
Add test file for experiments, and skeleton of custom UI for the plugin

tdesfont committed:
LeftBar custom params are working with simple angular

HenriChabert commented on resource/dku-helpers.js:
```diff
@@ -0,0 +1,93 @@
+
+function checkMandatoryParameters(param, webAppConfig) {
+    if (param.mandatory) {
+        var val = webAppConfig[param.name];
```
As said in the call, use `const` or `let` rather that `var`

tdesfont replied:
OK replaced mainly by let

tdesfont committed:
Add minimal example to recreate the UI for scatter plot charts in the geospatial toolkit density chart
leftBarParamsModule.js: Renamed, will contain the module for the left bar parameters
leftBarParamsTemplate.html: Renamed, will contain the template for the left bar parameters
webapp.json: Updated with the new module and template
Status: Display is working but need minor style adjustments, need to set the logic for scope variables transfer to graph

tdesfont committed:
Add mock top bar UI
topBarParamsModule.js: Will be the module handler for the top bar
topBarParamsTemplate.html: Template for the top bar module
webapp.json: Updated reference to the modules
Status: Visualisation still works due to the mock input hardcoded in the app.js file but the two paramsTemplate are disconnected from the logic for now

tdesfont committed:
Clarify the structure of the repository
- Moved html standalone tests in the tests/js
- Moved leaflet libs in resource/lib

tdesfont committed:
Include js libs (.gitignore libs)

tdesfont committed:
Revert to state without module or template
Trying to fix bug dataiku.webappBackend is undefined

tdesfont committed:
Fix bug for path of dku-helpers.js

tdesfont committed:
Minimal workaround for the custom UI, temporary and for testing purpose only
leftBarParamsTemplate.html: Set up a toy example for the UI
leftBarParamsModule.js

tdesfont committed:
Add filtering from graph-analytics, MVP for the tooltip
leftBarParamsModule.js: Correct names for parameters
filtering.py: Import module for filtering from graph-analytics
leftBarParamsTemplate.html: Toy responsive UI for parameters
app.js: Add more parameters handling
backend.py: Handle filtering and mock tooltip handling
body.html: Remove the slider in iframe (now unused)
Status: Working with responsive UI but no dataiku UI homogeneous, no OOP yet and need parameter on the tilemap option

tdesfont committed:
Refactoring and introduction of a main class to create a chartHandler singleton
app.js: Moved logic to create an intermediate chart layer between leaflet and dataiku custom chart
Status: Working with responsive UI but no dataiku UI homogeneous, need modification of the tilemap to not use mapbox token but carto

HenriChabert commented on js/leftBarParamsModule.js:
```diff
@@ -0,0 +1,13 @@
+
+app.controller('MyCustomFormController', ['$scope', function($scope) {
+    console.log("Calling the controller");
+    $scope.config.intensity = 50;
+    $scope.config.radius = 50;
+    $scope.config.color = {
+        name: 'blue'
+    };
+    $scope.config.maptile = {
+        name: 'mapbox/v11'
+    };
```
I would place the default values into constants (for example: `const DEFAULT_INTENSITY = 50`)

tdesfont replied:
Good idea, will do

HenriChabert commented on python-lib/dku_filtering/filtering.py:
```diff
@@ -0,0 +1,102 @@
+from functools import reduce
+
+
+def numerical_filter(df, filter):
```
name.
```suggestion
def numerical_filter(df, filter_):
```
If possible, it is better not to use builtin names in functions. The general workaround is to use var name `filter_`, that is the main role of placing an underscore at the end of a var

tdesfont replied:
Agreed

HenriChabert commented on python-lib/dku_filtering/filtering.py:
```diff
@@ -0,0 +1,102 @@
+    conditions = []
+    excluded_values = []
+    for k, v in filter['excludedValues'].items():
+        if k != '___dku_no_value___':
```
It would be better to use a constant DKU_NO_VALUE for that string.

HenriChabert commented on python-lib/dku_filtering/filtering.py:
```diff
@@ -0,0 +1,102 @@
+                excluded_values += [k]
+        else:
+            if v:
+                conditions += [~df[filter['column']].isnull()]
```
Is there a reason you are using a dict for `excludedValues` while you are only using it if the value is True ? Couldn't you just use a list with excluded values?

HenriChabert commented on python-lib/dku_filtering/filtering.py:
```diff
@@ -0,0 +1,102 @@
+def alphanum_filter(df, filter):
+    conditions = []
+    excluded_values = []
+    for k, v in filter['excludedValues'].items():
```
nitpicking: If using the var only in a Python script, prefer snake_case to remain consistent.

HenriChabert commented on python-lib/dku_filtering/filtering.py:
```diff
@@ -0,0 +1,102 @@
+    """
+    return a function to apply filtering conditions on df
+    """
+    if len(conditions) == 0:
```
nitpick: More pythonic to use `if not conditions`

HenriChabert commented on python-lib/dku_filtering/filtering.py:
```diff
@@ -0,0 +1,102 @@
+    elif len(conditions) == 1:
+        return df[conditions[0]]
+    else:
+        return df[reduce(lambda c1, c2: c1 & c2, conditions[1:], conditions[0])]
```
Nice way to do it!

HenriChabert commented on resource/dku-helpers.js:
```diff
@@ -0,0 +1,93 @@
+function checkMandatoryParameters(param, webAppConfig) {
+    if (param.mandatory) {
+        var val = webAppConfig[param.name];
+        if (val == undefined || val == "") {
```
```suggestion
        if (val === null || val === "") {
```
null returns true for both undefined and null, which is what you want unless you want null to be considered as ok.
You should also use `===` syntax that will run a deep check (verify type and value). For example, 1 == '1' returns true while 1 === '1' returns false. Using `===` makes you totally control what you want to check.

tdesfont replied:
Noted

HenriChabert commented on resource/dku-helpers.js:
```diff
@@ -0,0 +1,93 @@
+
+function checkWebAppParameters(webAppConfig, webAppDesc) {
+    if (webAppDesc.topBarParams) {
+        webAppDesc.topBarParams.forEach(p => {checkMandatoryParameters(p, webAppConfig)});
```
nitpick: When method has only one instruction, you do not need to use `{...}`.

HenriChabert commented on resource/dku-helpers.js:
```diff
@@ -0,0 +1,93 @@
+    } else {
+        var scale = 1 / (max - min);
+        return Math.pow(Math.max(0,(value - min)*scale), 2);
+    }
```
You do not need to use `else`. If condition in if verified, the function will return and the following instructions won't be executed.
```suggestion
    if (max === min) return 0;
    var scale = 1 / (max - min);
    return Math.pow(Math.max(0,(value - min)*scale), 2);
```

HenriChabert commented on resource/dku-helpers.js:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    function dkuDisplayError(error) {
+        console.warn("backend error: ", error);
```
Why using `console.warn()` insted of `console.error()` ?

tdesfont replied:
Will do

HenriChabert commented on resource/dku-helpers.js:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    function get(path, args={}, displayErrors=true) {
+        return fetch(getUrl(path) + '?' + $.param(args), {
```
```suggestion
        return fetch(`${getUrl(path)}?${$.param(args)}`, {
```
nitpick: I think that this syntax is most of the time clearer

HenriChabert commented on resource/dku-helpers.js:
```diff
@@ -0,0 +1,93 @@
+                if (response.status == 502) {
+                    throw Error("Webapp backend not started");
+                } else if (!response.ok) {
+                    response.text().then(text => dataiku.webappMessages.displayFatalError(`Backend error:\n${text}.\nCheck backend log for more information.`))
```
Indeed for the backend start. Why don' t you put it in app.js?

tdesfont replied:
What should we put in app.js ?

HenriChabert commented on resource/leftBarParamsTemplate.html:
```diff
@@ -0,0 +1,137 @@
+
+
+<!--
+&lt;!&ndash;TODO: Potentially leave only this &ndash;&gt;
```
I have tried before to just copy/paste the code of DSS and it has never worked. I think it is better to understand and to replicate only what you need.
About the dku-slider, it could be a great idea. Have you already tried using it ?
About the rest, it would indeed be good to replicate DKU style but it is already pretty nice I think.

HenriChabert commented on webapps/density-custom-chart/webapp.json:
```diff
@@ -0,0 +1,39 @@
+{
+    "meta": {
+        "label": "Geospatial density chart",
+        "description": "Custom chart to visualize datasets as graphs. These datasets must have source and target columns.",
```
Did you forget to modify it when you used the code of Stan?

tdesfont replied:
Yes thanks

HenriChabert commented on webapps/density-custom-chart/webapp.json:
```diff
@@ -0,0 +1,39 @@
+    "meta": {
+        "label": "Geospatial density chart",
+        "description": "Custom chart to visualize datasets as graphs. These datasets must have source and target columns.",
+        "icon": "icon-link"
```
Would use a different icon.

HenriChabert commented on webapps/density-custom-chart/body.html:
```diff
@@ -0,0 +1,17 @@
+<link rel="stylesheet" href="https://unpkg.com/leaflet@1.7.1/dist/leaflet.css" integrity="sha512-xodZBNTC5n17Xt2atTPuE1HxjVMSvLVW9ocqUKLsCC5CXdbqCmblAshOMAS6/keqq/sMZMZ19scR4PsZChSR7A==" crossorigin=""/>
+<script src="https://unpkg.com/leaflet@1.7.1/dist/leaflet.js" integrity="sha512-XQoYMqMTK8LvdxXYG3nZ448hOEQiglfqkJs1NOQV44cWnUrBc8PkAOcXy20w0vlaXaVUearIOBhiXZ5V3ynxwA==" crossorigin=""></script>
+<script src="/plugins/geospatial-toolkit/resource/libjs/leaflet-heat.js"></script>
+<script type="text/javascript"></script>
```
Empty script

tdesfont replied:
Removed

HenriChabert commented on webapps/density-custom-chart/backend.py:
```diff
@@ -0,0 +1,56 @@
+        # Fetch dataset from name
+        # TODO: Secure call to API
+        df = dataiku.Dataset(dataset_name).get_dataframe(limit=100000)
+        if len(filters) > 0:  # apply filters to dataframe
```
```suggestion
        if filters:  # apply filters to dataframe
```

HenriChabert commented on webapps/density-custom-chart/backend.py:
```diff
@@ -0,0 +1,56 @@
+        tooltip = ['A sample tooltip' for i in range(len(df))]
+        coordinates = list(df[geopoint_column_name].values)
+        long = [eval(i[6:-1].split(' ')[0]) for i in coordinates]
+        lat = [eval(i[6:-1].split(' ')[1]) for i in coordinates]
```
```suggestion
        long_, lat = zip(*[list(map(float, re.findall("\d+\.\d+", x))) for x in l])
```
Regex is more self explainatory I find but both are good, it is only to present you another way to do. Good practice also to use `long_` as long exists in Python builtin methods

HenriChabert commented on webapps/density-custom-chart/app.js:
```diff
@@ -0,0 +1,356 @@
+    Object.defineProperty(this, 'foo', {
+        get: function(){ return _foo;},
+        set: function(val){ _foo = val;}
+    });
```
I guess this is a test?

tdesfont replied:
Removed

HenriChabert commented on webapps/density-custom-chart/app.js:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    this.initialiseMap = function(){
+        _mapPointer = L.map(_mapId).setView([48.8444529, 2.3718228], 5);
```
Use constants to set DEFAULT_CENTER and DEFAULT_ZOOM

tdesfont committed:
Externalised the geo density chart object in separate js file

HenriChabert reviewed (changes requested):
Nice work !
I did not add reviews on the files that we checked together by Zoom. I'll look at it once you changed and tried what we discussed of :)

tdesfont committed:
UI modification

tdesfont committed:
Fix UI error

tdesfont committed:
Added visualisation of nearest neighbors for a small radius on the map

tdesfont committed:
Change tooltip

tdesfont committed:
Angular template hack

tdesfont committed:
Removed builtin name filter in python function

tdesfont committed:
PR fixes and enhancement

tdesfont committed:
Tweak UI of angular of reused components

tdesfont committed:
Tweak UI of angular of reused components

tdesfont committed:
Handle full dataiku palette for density color scale

tdesfont committed:
Set background color of tooltip

tdesfont committed:
Wording

tdesfont committed:
Added draft of tooltip retrieval for the density chart

tdesfont committed:
Added tooltip template in the webapp

tdesfont committed:
Revised tooltip process

tdesfont committed:
Implement default tooltip for categorical variables

tdesfont committed:
Refactoring and major change to the event handling, introduced an intermediary class to detect necessary recomputing of the charts.

tdesfont committed:
Merge pull request #3 from dataiku/feature/density-chart-tooltip

Feature/density chart tooltip

tdesfont committed:
[ch59606] Rename custom chart to density map

tdesfont committed:
[ch59349] Added landing image background and potential warning message display

tdesfont committed:
Added unit tests, refactoring of the backend
app.js: Rename function fullUpdate to updateCoreData
backend.py: Refactor backend to a single function call
test_dataset_retrieval.py: Create an integration test
test_geodata.py: Remove any debugging
dku_data_processing: Created a single directory for the python lib functions

tdesfont committed:
Set the quadtree search to only closest neighbor

tdesfont committed:
Secure centerMap call

tdesfont committed:
Merge branch 'feature/density-chart' into feature/density-chart-background-img

tdesfont committed:
Display select geo data log information

tdesfont committed:
Draft of the error handling dynamic

tdesfont committed:
Merge branch 'feature/density-chart-rename-chart' into feature/density-chart

tdesfont committed:
Change icon

tdesfont committed:
Final commit before review

tdesfont committed:
Fix painful mistake or bug to show the background image. (Never set the style display of the map to none or error and freeze). Hack, never put the map style as none and instead put the map background over the map.

tdesfont committed:
Set standard fonts for the leftBarParamsTemplate.html

tdesfont committed:
Formatted the tooltip

tdesfont committed:
Better handling of the error

tdesfont committed:
Add optional maps, test a higher sampling rate

tdesfont committed:
Include maptile selection for standard cartodb maptiles

tdesfont committed:
Handle invalid detail column through error warning

tdesfont committed:
Set the dynamic of the spinner

tdesfont committed:
Removed unused comments

tdesfont committed:
Fixed the default option for the maptile to use

tdesfont committed:
Add background images from UX

tdesfont committed:
Fix the maptile init value

tdesfont committed:
Keep aspect ratio of background image

tdesfont committed:
Set background color to red if warning

tdesfont committed:
Fix a defect in the case geopointColumn is null, should not recompute full backend data

tdesfont committed:
Resize background image
